File: typescript/src/podcast_extractor_client.ts
export type PodcastDiscoverInput = {
  platform: string
  sourceUrl: string
  publishedAfter?: string
  limit?: number
}

export type PodcastExtractInput = {
  platform: string
  episodeUrl: string
}

export type PodcastExtractorRequestFn = (
  url: string,
  method: string,
  headers: Record<string, string>,
  body: string,
) => Promise<{ status: number; text: string }>

export type PodcastExtractorClient = {
  discoverEpisodes(input: PodcastDiscoverInput): Promise<Record<string, unknown>>
  extractAudio(input: PodcastExtractInput): Promise<Record<string, unknown>>
}

const normalizeBaseUrl = (raw: string): string => raw.trim().replace(/\/+$/, "")

const defaultRequest: PodcastExtractorRequestFn = async (url, method, headers, body) => {
  const resp = await fetch(url, { method, headers, body, signal: AbortSignal.timeout(20_000) })
  return { status: resp.status, text: await resp.text() }
}

const parseJSON = (text: string): Record<string, unknown> => {
  try {
    const parsed = JSON.parse(text)
    if (parsed && typeof parsed === "object") return parsed as Record<string, unknown>
  } catch {
    return {}
  }
  return {}
}

export const buildPodcastExtractorClient = (
  capabilityEndpoint: string,
  ephemeralToken: string,
  requestFn?: PodcastExtractorRequestFn,
): PodcastExtractorClient => {
  const base = normalizeBaseUrl(capabilityEndpoint)
  const token = ephemeralToken.trim()
  const request = requestFn ?? defaultRequest

  const post = async (route: string, payload: Record<string, unknown>): Promise<Record<string, unknown>> => {
    if (!base) throw new Error("podcast extractor endpoint is required")
    const headers: Record<string, string> = { "Content-Type": "application/json" }
    if (token) headers.Authorization = `Bearer ${token}`
    const { status, text } = await request(`${base}${route}`, "POST", headers, JSON.stringify(payload))
    if (status < 200 || status >= 300) throw new Error(`podcast extractor request failed: http ${status} ${text}`)
    const parsed = parseJSON(text)
    return (parsed.data ?? parsed) as Record<string, unknown>
  }

  return {
    discoverEpisodes: async (input) =>
      await post("/api/internal/podcasts/discover", {
        platform: input.platform,
        source_url: input.sourceUrl,
        published_after: (input.publishedAfter ?? "").trim(),
        limit: input.limit ?? 0,
      }),
    extractAudio: async (input) =>
      await post("/api/internal/podcasts/extract", {
        platform: input.platform,
        episode_url: input.episodeUrl,
      }),
  }
}

File: python/octoarms_sdk/podcast_extractor_client.py
from __future__ import annotations

import json
import urllib.error
import urllib.request
from collections.abc import Callable
from typing import Any


PodcastExtractorRequestFn = Callable[[str, str, dict[str, str], bytes], tuple[int, str]]


def _normalize_base_url(raw: str) -> str:
    return raw.strip().rstrip("/")


def _default_request(url: str, method: str, headers: dict[str, str], body: bytes) -> tuple[int, str]:
    req = urllib.request.Request(url=url, data=body, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=20) as response:
            status = getattr(response, "status", response.getcode())
            return int(status), response.read().decode("utf-8")
    except urllib.error.HTTPError as exc:
        return int(exc.code), exc.read().decode("utf-8", errors="replace")


def _parse_response(text: str) -> dict[str, Any]:
    try:
        parsed = json.loads(text)
    except json.JSONDecodeError:
        return {}
    if not isinstance(parsed, dict):
        return {}
    data = parsed.get("data")
    if isinstance(data, dict):
        return data
    return parsed


class PodcastExtractorClient:
    def __init__(self, *, base_url: str, token: str, request_fn: PodcastExtractorRequestFn) -> None:
        self._base_url = base_url
        self._token = token
        self._request = request_fn

    def _post(self, route: str, payload: dict[str, Any]) -> dict[str, Any]:
        if not self._base_url:
            raise RuntimeError("podcast extractor endpoint is required")

        headers = {"Content-Type": "application/json"}
        if self._token:
            headers["Authorization"] = f"Bearer {self._token}"

        status, text = self._request(
            f"{self._base_url}{route}",
            "POST",
            headers,
            json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        )
        if status < 200 or status >= 300:
            raise RuntimeError(f"podcast extractor request failed: http {status} {text}")
        return _parse_response(text)

    def discover_episodes(
        self,
        *,
        platform: str,
        source_url: str,
        published_after: str = "",
        limit: int = 0,
    ) -> dict[str, Any]:
        return self._post(
            "/api/internal/podcasts/discover",
            {
                "platform": str(platform),
                "source_url": str(source_url),
                "published_after": str(published_after).strip(),
                "limit": int(limit or 0),
            },
        )

    def extract_audio(self, *, platform: str, episode_url: str) -> dict[str, Any]:
        return self._post(
            "/api/internal/podcasts/extract",
            {
                "platform": str(platform),
                "episode_url": str(episode_url),
            },
        )


def build_podcast_extractor_client(
    capability_endpoint: str,
    ephemeral_token: str,
    request_fn: PodcastExtractorRequestFn | None = None,
) -> PodcastExtractorClient:
    return PodcastExtractorClient(
        base_url=_normalize_base_url(capability_endpoint),
        token=ephemeral_token.strip(),
        request_fn=request_fn or _default_request,
    )
